src/crl/interactivesessions/shells/remotemodules/msgcaches.py

```diff
@@ -0,0 +1,159 @@
+import time
+import itertools
+
+
+__copyright__ = 'Copyright (C) 2019, Nokia'
+
+
+class MsgCaches(object):
+    def __init__(self, retry, send_msg):
+        self._retry = retry
+        self._send_msg = send_msg
+        self._msgcaches = {}
+        self._monotonic = Monotonic()
+        self._lap = self._monotonic.time()
+        self._removed = ConsecutiveSet()
+
+    @property
+    def _elapsed(self):
+        return self._monotonic.time() - self._lap
+
+    def push_msg(self, msg):
+        self._msgcaches[msg.uid] = MsgCache(msg, self._retry, send_msg=self._send_msg)
+
+    def send_expired(self):
+        for uid, c in self._msgcaches.items():
+            c.send_expired(self._elapsed)
+            if c.timeout == Infinite():
+                self._remove(uid)
+        self._lap = self._monotonic.time()
+
+    def remove(self, msg):
+        if msg.uid in self._msgcaches:
+            self._remove(msg.uid)
+
+    def _remove(self, uid):
+        del self._msgcaches[uid]
+        self._removed.add(uid)
+
+    def get_msg(self, uid):
+        if uid in self._msgcaches:
+            return self._msgcaches[uid].msg
+        if uid in self._removed:
+            raise MsgCachesAlreadyRemoved()
+        else:
+            raise MsgCachesNotFound()
+
+    @property
+    def timeout(self):
+        return min(itertools.chain(self._timeouts_gen(),
+                                   iter([Infinite()])))
+
+    @property
+    def timeout_args(self):
+        t = self.timeout
+        return [] if t == Infinite() else [t]
+
+    def _timeouts_gen(self):
+        for _, c in self._msgcaches.items():
+            yield c.timeout
+
+    @property
+    def msgs(self):
+        return list(self._msg_gen())
+
+    def _msg_gen(self):
+        for _, c in self._msgcaches.items():
+            for m in c.msgs:
+                yield m
+
+
+class MsgCachesAlreadyRemoved(Exception):
+    pass
+
+
+class MsgCachesNotFound(Exception):
+    pass
+
+
+class ConsecutiveSet(object):
+    """ Set for storing efficiently mostly consecutive integers.
+    """
+    def __init__(self):
+        self._min = None
+        self._max = None
+        self._missing_nbrs = set()
+
+    def add(self, i):
+        if self._min is None:
+            self._min = self._max = i
+        elif i > self._max:
+            self._missing_nbrs |= set(range(self._max + 1, i))
+            self._max = i
+        elif self._min < i < self._max and i in self._missing_nbrs:
+            self._missing_nbrs.remove(i)
+        elif i < self._min:
+            self._missing_nbrs |= set(range(i + 1, self._min))
+            self._min = i
+
+    def __contains__(self, i):
+        return self._min <= i and i <= self._max and i not in self._missing_nbrs
```

Comment:
Can be made shorter    `self._min <= i and i <= self._max`--> self._min <= i <= self._max